controller/system_rollout_controller.go

```diff
@@ -0,0 +1,1801 @@
+		lock: &sync.Mutex{},
+	}
+
+	return c
+}
+
+func (c *SystemRolloutController) controlleeHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueControlleeChange() },
+		UpdateFunc: func(old, cur interface{}) { c.enqueueControlleeChange() },
+		DeleteFunc: func(obj interface{}) { c.enqueueControlleeChange() },
+	}
+}
+
+func (c *SystemRolloutController) namespacedControlleeHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueNamespacedControlleeChange(obj) },
+		UpdateFunc: func(old, cur interface{}) { c.enqueueNamespacedControlleeChange(cur) },
+		DeleteFunc: func(obj interface{}) { c.enqueueNamespacedControlleeChange(obj) },
+	}
+}
+
+func (c *SystemRolloutController) enqueue() {
+	c.queue.Add("system-rollout")
+}
+
+func (c *SystemRolloutController) enqueueControlleeChange() {
+	c.queue.Add("controllee")
+}
+
+func (c *SystemRolloutController) enqueueNamespacedControlleeChange(obj interface{}) {
+	metadata, err := meta.Accessor(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get meta for object %#v: %v", obj, err))
+		return
+	}
+
+	if metadata.GetNamespace() == os.Getenv(EnvLonghornNamespace) {
+		c.enqueueControlleeChange()
+	}
+}
+
+func (c *SystemRolloutController) Run() error {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	if !cache.WaitForNamedCacheSync("longhorn system rollout", c.stopCh, c.cacheSyncs...) {
+		return fmt.Errorf("failed to sync informers")
+	}
+
+	startTime := time.Now()
+	defer func() {
+		log := c.logger.WithField("runtime", time.Since(startTime))
+		log.Info("Closing")
+	}()
+	go wait.Until(c.worker, time.Second, c.stopCh)
+
+	<-c.stopCh
+	return nil
+}
+
+func (c *SystemRolloutController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *SystemRolloutController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	defer c.cacheStateReset()
+
+	err := c.syncSystemRollout()
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *SystemRolloutController) syncSystemRollout() error {
+	err := c.syncController()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		log := c.getLoggerForSystemRollout()
+		systemRestore, err := c.ds.GetSystemRestore(c.systemRestoreName)
+		if err != nil {
+			log.WithError(err).Error("failed to get SystemRestore")
+			close(c.stopCh)
+		}
+
+		notCompleted := systemRestore.Status.State != longhorn.SystemRestoreStateCompleted
+		notFailed := systemRestore.Status.State != longhorn.SystemRestoreStateError
+		if notCompleted && notFailed {
+			log.Infof(SystemRolloutMsgEnqueueFmt, systemRestore.Status.State)
+			c.enqueue()
+			return
+		}
+
+		_, err = c.ds.RemoveSystemRestoreLabel(systemRestore)
+		if err != nil {
+			log.WithError(err).Warnf(SystemRolloutMsgEnqueueFmt, "failed to remove SystemRestore label")
```

Comment:
`"failed to remove SystemRestore label after the restore state %v"`

Reply:
Updated.